GSDX : Add an option for CRTC size and some other video mode stuffs

ssakash opened the pull request:
**Summary of changes:**

~~* Add an option for customizing the CRTC size of game to standard NTSC/PAL variants. ( deals with all the remaining [PCRTC issues](http://forums.pcsx2.net/Forum-Bug-reporting?sortby=lastpost&order=desc&datecut=9999&prefix=13) )~~

~~**Note:** Currently I've added it at the place of OpenCL device combobox which is currently removed from normal release builds unless you define a certain macro, if you define the macro on code then CRTC size combobox won't be available. for future - It'd be better to make tab controls similar to Linux GUI and sort all the options in a neat order.~~

![dialog](http://ultraimg.com/images/Screenshot148f451.png)
~~>Fixes black border issue on King of Fighters 2000 when CRTC size is set to 640 X 224 (NTSC-P) , here's a comparison picture for relevance.~~

**Old behavior:**

![broken KOF](http://ultraimg.com/images/gsdx_20160305103535.png)

**CRTC size : 640 X 224 (NTSC-P)**

![fixed KOF](http://ultraimg.com/images/gsdx_20160305103524.png)

~~>Strangely our current height auto detection fails only for the following game , the game even allows pixels to be drawn till the black border part according to the SCAY1 value of the scissoring test but they seem to remain black the whole time whereas natively on the PS2 you can't reproduce the black borders.~~

~~>The following change is also useful for people with patched binaries. there are cases where people patch their games to make use of different video modes. example: NTSC games using PAL video mode. since currently we rely heavily on the video mode for proper height detection , some patched games which make use of non-native video modes might have some conflicts. the following CRTC size combobox might also be useful at such scenarios :)~~

* EE-SYSCALL: Add some missing video modes and make the console messages be printed even without checking Dev/Verbose option on the console window.

>The console messages are useful for identifying the video modes and when users ask support for issues with patched binaries , we can easily identify such cases by taking a look at emulog file instead of asking them to enable Dev/verbose option all the time.

willkuer commented:
If i understand correctly this tackles problems that happen very seldomly? Wouldn't it be nicer to keep the option hidden in the ini file then to reduce confusion of new users?

I would always recommend enums over integers. The modes are declared at one point during combobox initialization and handled at a different point in a switch. If you now need ro adjust the switch you have to jump back and forth to understand the int values. Imo enums/enum classes would be better.

ssakash commented:
> If i understand correctly this tackles problems that happen very seldomly? 

The combobox option will also be useful for people with patched binaries, I'm not sure whether it's a rare case.

> Wouldn't it be nicer to keep the option hidden in the ini file then to reduce confusion of new users?

Initially I also had thought of something like that but unfortunately , this is the only method to solve issues with patched binaries. (see my first post) and keeping it hidden will just make it hidden and a trouble for other users to use different options of the combobox frequently. It'd be better to add a nice tooltip to prevent user confusion.

> I would always recommend enums over integers. The modes are declared at one point during combobox initialization and handled at a different point in a switch. If you now need ro adjust the switch you have to jump back and forth to understand the int values. Imo enums/enum classes would be better.

I don't really think enums are ideal here. it also saves me some casting ;)

FlatOutPS2 commented:
I'd rather see something like that provided in HW Hacks, it would cause less confusion.

The Plugin provided here: http://forums.pcsx2.net/Thread-GSDX-PCRTC-Bug-Report-The-King-of-Fighters-2000-SLPS-251-56?page=6 also fixed a similar issue in Metal Slug 3 and Kengo: Master Of Bushido.

turtleli commented:
Hmm. I don't see this option being used much, and changing it to anything but "Auto" would likely cause issues. So I also think it would be better placed in hacks, which probably requires shuffling GUI stuff around a bit since it's not HW only.

As for the enums, I also think it would be a good idea. You shouldn't need to cast too much - willkuer did stuff in #1018 that should cut down on the amount of casts that you need.

turtleli commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -131,6 +131,8 @@ const char* dialog_message(int ID, bool* updateText) {
 		case IDC_MIPMAP:
 			return "Enables mipmapping, which some games require to render correctly. Turn off only for debug purposes.";
+		case IDC_CRTC_SIZE:
+			return "Allows configuration of the display rectangle to other standard PAL/NTSC variants. \nNote: NTSC-P/PAL-P option only works during progressive scan modes whereas NTSC/PAL works at interlaced video modes.";
```
Spaces unnecessary before `\n`.

turtleli commented on plugins/GSdx/GSState.cpp:
```diff
@@ -351,7 +352,48 @@ GSVector4i GSState::GetDisplayRect(int i)
+	case 1:
+		if (Interlaced_mode)
+		{
+			width = 640; height = 448;
```
I prefer not having two statements on one line.

turtleli commented on plugins/GSdx/GSdx.rc:
```diff
@@ -209,6 +209,8 @@ BEGIN
     LTEXT           "OpenCL Device:",IDC_OPENCL_TEXT,6,102,53,8
     COMBOBOX        IDC_OPENCL_DEVICE,71,100,166,118,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
+    COMBOBOX        IDC_CRTC_SIZE, 71, 100, 166, 118, CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
+    LTEXT           "CRTC Size:", IDC_CRTC_SIZE_TEXT, 6, 102, 53, 8
```
Hand-edited? Compare the space after the commas with the other lines.

gregory38 commented:
```
 +enum class GSCRTC_Size : int8_t
+{
+	Auto,
+	NTSC_640X448,
+	NTSC_512X448,
+	NTSC_PROGRESSIVE_640X224,
+	PAL_640X512,
+	PAL_512X512,
+	PAL_PROGRESSIVE_640X256
+};
```
Do we really need all those possibilities? Progressive autodetection is always correct isn't it? Same for the width?

refractionpcsx2 commented:
If it only happens on one game, I wouldn't like to see an interface option added for it, this is a bit of a waste of space, it would be more beneficial to have it as a CRC hack or something.

ssakash commented:
Added enums to code , thanks to turtleli and willkuer for the suggestion.

> Do we really need all those possibilities? Progressive autodetection is always correct isn't it? Same for the width?

I have added some of the standard NTSC/PAL resolutions. only the progressive mode height auto detection is wrong for the following game , interlaced mode is fine. You're right, currently we don't have any issues with width detection on any games but I also added them into the option since it'd seem a bit weird to only include an option to change the CRTC height.

> If it only happens on one game, I wouldn't like to see an interface option added for it, this is a bit of a waste of space, it would be more beneficial to have it as a CRC hack or something.

you are forgetting about people with patched binaries :P 

though, generally I'm not in favor of CRC hacks since they hide the problem. so even if you fix the problem accurately it takes a bit of time to notice it.

also back to the topic on placement of the option : as turtleli said if we want the option on HW hacks dialog , we need to do some shuffling and probably change it into uhm.. "Hacks dialog" since CRTC size also influences the software renderer. another option would be to only make the option influence hardware mode and place it at HW hacks. opinions ?

ssakash commented:
Also I made a dedicated option for only influencing the progressive mode since progressive mode usually uses half height compared to the interlaced mode so if you force the bigger height value on progressive mode then it'll screw up the display.

FlatOutPS2 commented:
> also back to the topic on placement of the option : as turtleli said if we want the option on HW hacks dialog , we need to do some shuffling and probably change it into uhm.. "Hacks dialog" since CRTC size also influences the software renderer. another option would be to only make the option influence hardware mode and place it at HW hacks. opinions ?

No, it has to also work in software mode. The games affected are mostly 2D games that work better in software mode.

ssakash commented:
> No, it should also work in software mode. The games affected are mostly 2D games that work better in software mode.

That's a nice point. I'll just remodel the hacks dialog then.

gregory38 commented:
I'm in favor to reduce the possibilities of the combobox to the minimal.

ssakash commented:
updated the hacks dialog and some minor refactoring to the main GSDX dialog.

![hacks_dialog](http://ultraimg.com/images/Screenshot1966096.png)

![main_dialog](http://ultraimg.com/images/Screenshot16afac5.png)

**Summary of new changes:**

* **Enable Hacks** checkbox was removed for consistency similar to Shader configuration. please also refer the [third commit](https://github.com/PCSX2-Alpha/pcsx2/commit/c29e3629a53cf1eddcef800e79a1c110fb3ec189). maybe we could add a restore defaults option on the hacks dialog if users need to change to the default values but I rather doubt it's necessary since the default values are practically all the options turned off. @gregory38 : what's your opinion on this ?

* Hacks for upscaling get grayed out on native resolution.

* _**Note:**_ Skipdraw , TCOffset text containers don't have the same width as the combobox at hacks dialog since they're different dialog elements. (compare with rendering threads text container on main dialog)

@turtleli : Please review the code :)

ssakash commented:
> Correct me, if I'm wrong. Current detection code is working to detect interlaced/progressive.
Manual selection will be a PITA if the game toggle between the 2 modes.

The reason I have progressive mode exclusive options is due to the fact that some games only get the height for progressive mode _wrong_. the interlaced width/height values are correct. even if the game constantly changes between Interlaced/progressive mode when a progressive mode option is selected there should be no problem, we'll use the autodetection in that case when interlaced mode is used so the game should be fine.

The progressive mode exclusive options are only intended for games such as KOF which have issues with progressive mode height detection. ( currently, we don't have any other cases of games other than KOF )

>  So in my opinion, the progressive options must be removed and automatically detected.

I think you want the interlaced options to also handle the progressive mode cases ? I don't think that will work well.

KOF uses ``640X512`` for **Interlaced** (display is fine) but at **progressive mode** it uses ``640X256`` ( [black borders are present on both NTSC/PAL Vmodes](https://camo.githubusercontent.com/e7ab9eb120daf02f42efdaab83620f12f7d964be/687474703a2f2f756c747261696d672e636f6d2f696d616765732f677364785f32303136303330353130333533352e706e67) ) , ``640X224`` ( [black borders are gone on both NTSC/PAL Vmodes](https://camo.githubusercontent.com/f3856caa9d1ba4de8570da44bedb6e50da5a0424/687474703a2f2f756c747261696d672e636f6d2f696d616765732f677364785f32303136303330353130333532342e706e67)).

So if you want an interlaced option to handle the progressive mode detection for KOF then you'll need ``640X448`` to achieve ``640X224`` on progressive mode. but the game in question (KOF) needs ``640X512`` for proper display on interlaced modes hence I have options exclusive to progressive mode.

> To be discussed, width is also properly detected so only the number of lines is useful. But I'm open.

To be honest, the only reason I keep the width is to have a consistency on the GUI. Here's an example:

**Combobox (With width):**
* 640 X 448 (NTSC)
* 512 X 448 (NTSC)
* 640 X 224 (NTSC-P)
* 640 X 512 (PAL)
* 512 X 512 (PAL)
* 640 X 256 (PAL-P)

``Remarks : looks quite nice since we mention the standard resolutions for video modes.``

**Combobox (Without width):**
* 448 (NTSC)
* 224 (NTSC-P)
* 512 (PAL)
* 256 (PAL-P)

``Remarks : not so nice ! we only include the height with respect to the video mode which seems a bit weird``

> Note: I think you want to alter the dimension only if game is PAL or NTSC.

you mean PAL option should only be selection by PAL video modes ? the game in question - KOF needs 640X224 (NTSC-P) for both the PAL/NTSC video modes to work without the black borders.

> The goal was to easily disable, all HW hack? Now you need to remember the default of the N options.

Yes, that was aim of the userhacks option, the default for options shouldn't be a problem since the default value for all HW hacks option is 0. ( Restore defaults box could probably be added but I rather doubt it's necessary, we also don't have any checkbox for enabling the use of shader dialog options (Shadeboost, FXAA), so I think this makes the behavior more consistent)

@willkuer @karasuhebi @FlatOutPS2 @refractionpcsx2  - If you don't mind, can you provide feedback on the [new change to the GSDX dialog](https://github.com/PCSX2/pcsx2/pull/1217#issuecomment-194761303) , I think the GUI could be organized better.

also does someone have any nice idea to deal with the ``Shader/Hacks Configuration`` -> ``Configure`` redundancy ?

turtleli commented:
Do any games get the interlaced settings wrong?

I think you should keep the enable hacks checkbox instead of deleting it, it's useful as a quick disable switch. Maybe move it to the Hacks dialog if you want consistency.

> also does someone have any nice idea to deal with the Shader/Hacks Configuration -> Configure redundancy ?

"Configure Shaders" and "Configure Hacks" buttons with no accompanying label?

FlatOutPS2 commented:
> I think you should keep the enable hacks checkbox instead of deleting it, it's useful as a quick disable switch.
I was just thinking the same, but not about moving the button to the hacks dialog, that would make it useless.
It's already a hassle to get to the video plugin settings menu each time for debugging, needing to go to the hacks config each time to switch it off would make it as much work as turning off individual hacks by themselves.
I also preferred the Hacks config to be in the Hardware Mode Settings area, most hacks don't really apply to the software mode, so it at least be made clearer for the average user.

willkuer commented:
One could order hacks in the hacks dialog into groups of HW, SW and global. However if you put more into the hacks dialog it easily gets larger than the main dialog and with that also more confusing.

Also it is not entirely clear why some elements are in the hacks section while others are in the main dialog. Best example is maybe the crc hacks combobox. Obviously it provides hacks. The same for SW AA1. SW Mipmapping, preload data frame and hardware depth are rather inverted hacks. Disabling them is a hack and not a setting.

Don't know how one can make thinks easier. It could be good to have a user mode and an advanced mode. User mode only allows you to select renderer and internal resolution and advanced allows you to adjust everything. If one does it like this the advanced settings can be just a table/listview of settings and the user mode will be extremely user friendly (with something like 2-3 settings).

gregory38 commented:
If nobody reports issue with the fast accurate blending. I think later we could remove the option from the GUI. The "depth emulation option" will later be moved as a speed hack.

The idea of CRC is to have an easy control without enabling all others hacks.

```
I think you want the interlaced options to also handle the progressive mode cases ? I don't think that will work well.

KOF uses 640X512 for Interlaced (display is fine) but at progressive mode it uses 640X256 ( black borders are present on both NTSC/PAL Vmodes ) , 640X224 ( black borders are gone on both NTSC/PAL Vmodes).
```
I'm a bit lost ;) What happen if KOF uses interlaced for some frames and progressive for others frames.

gregory38 commented:
Maybe you can make a separate PR with the first, and the 2 last commit.

Otherwise:
* 2nd commit must be split. GUI change | New CRTC option (Well if people agree with the GUI)
* 3rd commit is no. Users are lazy and it would just create more supports for no gain
*4th commit, is still worth it to do the saturation? Now that we have the option. Basically we need to chose between a cut-screen or a black border.

Note for the future: a mechanism to register the size on the ini (CRC). This way users don't need to toggle the options every games.

ssakash commented:
I removed the GUI code for CRTC size combobox. I'm planning on fixing it accurately by delving into these undocumented GS registers !

gregory38 commented:
I'm fine with the code. @turtleli is it ok for you?

turtleli commented:
Yeah, code seems ok.
